lib/kcerts/certs.go

```diff
@@ -0,0 +1,204 @@
+func GenerateServerKey(opts *CertOptions, intermediateCert *x509.Certificate, intermediatePrivateKey *rsa.PrivateKey) (*x509.Certificate, []byte, *rsa.PrivateKey, error) {
+	serverPrivateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	if !opts.isValid {
+		return nil, nil, nil, errors.New("must call Validate() on certificate options before creating a Server Crt")
+	}
+
+	serverTemplate := &x509.Certificate{
+		SerialNumber:   big.NewInt(1),
+		NotBefore:      opts.Before,
+		NotAfter:       opts.After,
+		KeyUsage:       x509.KeyUsageCRLSign,
+		ExtKeyUsage:    []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		IsCA:           false,
+		MaxPathLenZero: true,
+		IPAddresses:    opts.IPAddresses,
+	}
+
+	certBytes, err := x509.CreateCertificate(rand.Reader, serverTemplate, intermediateCert, &serverPrivateKey.PublicKey, intermediatePrivateKey)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	cert, err := x509.ParseCertificate(certBytes)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	b := pem.Block{Type: "CERTIFICATE", Bytes: certBytes}
+	certPEM := pem.EncodeToMemory(&b)
+
+	return cert, certPEM, serverPrivateKey, nil
+
+}
+
+// GenerateSSHKeyPair will generate a re-encoded private rsa key and public key from an existing *rsa.PrivateKey.
+func GenerateSSHKeyPair(privateKey *rsa.PrivateKey) ([]byte, []byte) {
+	privateBlock := &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
+	}
+
+	publicBlock := &pem.Block{
+		Type:  "PUBLIC KEY",
+		Bytes: x509.MarshalPKCS1PublicKey(&privateKey.PublicKey),
+	}
+	return pem.EncodeToMemory(publicBlock), pem.EncodeToMemory(privateBlock)
+}
+
+type CertOptions struct {
+	Country      []string
+	Organization []string
+	Before       time.Time
+	After        time.Time
+	IPAddresses  []net.IP
+	isValid      bool
+}
+
+func (o *CertOptions) WithCountries(cs []string) *CertOptions {
+	o.Country = cs
+	return o
+}
+
+func (o *CertOptions) WithOrganizations(orgs []string) *CertOptions {
+	o.Organization = orgs
+	return o
+}
+
+func (o *CertOptions) WithIpAddresses(ips []net.IP) *CertOptions {
+	o.IPAddresses = ips
+	return o
+}
+
+func (o *CertOptions) ValidUntil(time2 time.Time) *CertOptions {
+	o.After = time2
+	return o
+}
+func (o *CertOptions) NotValidBefore(time2 time.Time) *CertOptions {
+	o.Before = time2
+	return o
+}
+
+func (o *CertOptions) Validate() error {
+	currTime := time.Now()
+	if currTime.Before(o.Before) {
+		return errors.New("time must be before time.Now")
+	}
+	if currTime.After(o.After) {
+		return errors.New("cannot issue invalid CA's time invalid")
+	}
+	if o.After.Sub(currTime).Hours() < 24*365 { // hours in a year
+		return errors.New("duration of the CA is too low")
+	}
+	if len(o.Organization) == 0 {
+		return errors.New("must set organization")
+	}
+	if len(o.Country) == 0 {
+		return errors.New("must set countries of origin for en_Lang and i8n support")
+	}
+	o.isValid = true
```

Comment:
(optional: I'd be ok with merging as is, and iterate later - let me know your thoughts)

I'm not a big fan of o.isValid. The reason is that 1) user of the API must remember to call it, and 2) parameters can be changed after validation, accidentally, and 3) it's tricky to provide optional computed parameters (less relevant here, but somewhat). Eg:

```
opts := NewOptions()
opts.Before = ...
opts.After = ...
opts.... =

opts.Validate()
opts.Before = ...
``` 
I've not been consistent either in the rest of the code base, but I'm a fan of the "Functional Options" pattern (google it, a lot of good presentations and talks), that can prevent errors here and allow computed defaults. 

This is how your code would look like:
```
// Attributes are private to this package.
// Could even mark the structure private? Up to you.
type CertOptions struct {
	country      []string
	organization []string
	before       time.Time
	after        time.Time
	ipAddresses  []net.IP
}

type Modifier func (*CertOptions) error

func WithCountries(cs []string) Modifier {
	return func (o *CertOptions) error {
          o.Country = cs
          return nil
	}
}


func NewOptions(mods ...Modifier) (*CertOptions, error) {
  opts := &CertOptions{
    Before: time.Now(), // Some defaults if necessary
  }
  for _, mod := range mods {
    if err := mod(opts); err != nil {
      return nil, err
    }
  }
  // Update defaults based on other defaults, if necessary.
  // eg, bad idea here, but if opts.After == 0 { opts.After = opts.Before + 7 * 24 * time.Hour }, default 7 days validity.
  if err := Validate(opts); err != nil {
    return nil, err
  }
  return opts, nil
}
```

This way, it is impossible to generate an invalid CertOptions, and no need to check for validation further down.
It also allows to provide reasonable default in the API.

If parameters are mandatory, just make them as mandatory arguments for the `NewOptions` function, so build time failure rather than run time.

Reply:
The reason why I generally prefer the `func(r *Reciever) *Reciever` configuration method is actually for onboarding/devtooling.

Mainly, when I hit `kcerts.NewOptions().` the period would autocomplete in vim/vscode/idea etc and give the developer all the options available without having to trawl through the original code. 

I kinda screwed up not having the exported funcs use validate. I should not be using isValid under any circumstance and it's really bad smell. 

I'm fine using the functional modifiers, I find it incredibly irritating that even in the grpc and google-client package they use both methodologies.

Reply:
If you're not swayed by the tooling argument I'll just convert to the functional modifiers, either way is standard practice and I have no issues with using it.